### PR TITLE
[Sherwin Ee] iP

### PR DESCRIPTION
**Pull request by sherwinee**
# Dash [🤖](https://emojipedia.org/robot)

> "The secret of getting ahead is getting started. The secret of getting started is breaking your complex overwhelming tasks into small manageable tasks, and then starting on the first one." — Mark Twain ([source](https://chatgpt.com/share/67b1b0e7-5e44-8004-ae43-6913b80a0b5a))

Dash is a **FREE** chatbot that helps you keep track of tasks and events when you're too busy to remember everything! Its:
- Typing based
- **Fast** and **Easy to Use**
- Uniquely Singaporean (_it speaks like one_)

To use it, just 
1. Download it
2. Open it
3. Talk to it

## What it can do
- [ ] Keep track of Todos
- [ ] Keep track of Deadlines
- [ ] Keep track of Events

## Usage
`List` - Lists all tasks
`Search <String>` - Search for tasks containing the given string.
`todo <Task Name>` - Adds a Todo task
`deadline <Task Name> /by <Deadline YYYY-MM-DD>` - Adds a Deadline task
`event <Task Name> /from <Start Date YYYY-MM-DD> /to <End Date YYYY-MM-DD>` - Adds an Event
`delete <Index>` - Deletes task at given index
`deleteall `- Clears entire task list

Change the bot's name in the `Dash` class:
```java
public class Dash {
    public static final String botName = "Dash"; // Change this if you wish!

    private static final String filePath = "./data/dash.txt";
    private final Ui ui;
    private final Storage storage;
    private final TaskList taskList;
    private final Parser parser;
    ...
```

### CODE REVIEW

File: src/main/java/Command.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public enum Command {
```
Comment (0x4F776C) [inline src/main/java/Command.java]: Could we add JavaDoc comments for this class?

File: src/main/java/Dash.java [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    private static final String FILE_PATH = "./data/dash.txt";
+
+    public static String getTaskListString() {
+        return taskList.stream()
+                .map(task -> task.stringify() + "\n")
+                .reduce("", (x, y) -> x + y)
+                .strip();
+    }
+
+    private static Task getTaskFromString(String str) throws IllegalArgumentException {
```
Comment (0x4F776C) [inline src/main/java/Dash.java]: Could we add JavaDoc comments for this method, including @param and @return tags?

File: src/main/java/Dash.java [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+            task.markDone();
+            botAddLine("Ok! I mark this task as done liao!");
+            botAddLine("  " + task);
+            botPrint();
+        } catch (NumberFormatException e) {
+            botAddLine("Invalid number format!");
+            botPrint();
+        } catch (IndexOutOfBoundsException e) {
+            botAddLine("That task doesn't exist!");
+            botPrint();
```
Comment (0x4F776C) [inline src/main/java/Dash.java]: This block of code appears in multiple places. Should we extract it into a separate method to avoid duplication?

File: src/main/java/Dash.java [hunk 4]
```diff
@@ -0,0 +1,330 @@
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.*;
```
Comment (0x4F776C) [inline src/main/java/Dash.java]: Should we remove the implicit import to keep them explicit?

File: src/main/java/Dash.java [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+            case EVENT:
+                addEvent(msg);
+                break;
+
+            default:
+                printDefaultMessage();
+                break;
+            }
+            }
```
Comment (0x4F776C) [inline src/main/java/Dash.java]: Could we ensure consistent indentation here?

File: text-ui-test/EXPECTED.TXT [hunk 6]
```diff
@@ -1,7 +1,68 @@
...
-|  _ \ _   _| | _____ 
-| | | | | | | |/ / _ \
-| |_| | |_| |   <  __/
-|____/ \__,_|_|\_\___|
+    ____________________________________________________________
+     Hello! I'm Dash
+     What you want me do today?
+    ____________________________________________________________
 
+    ____________________________________________________________
```
Comment (0x4F776C) [inline text-ui-test/EXPECTED.TXT:10]: Could we add test cases for edge scenarios, like empty input or invalid values?

**Review by 0x4F776C — COMMENTED** [review]
LGTM! However, OOP-structure should be integrated as **Dash.java** seems to be performing all the tasks.

File: src/main/java/Command.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    MARK("mark"),
+    UNMARK("unmark"),
+    DELETE("delete", "del"),
+    DELETEALL("deleteall", "delall"),
+    TODO("todo"),
+    DEADLINE("deadline"),
+    EVENT("event");
+
+    private static final Map<String, Command> commandMap = new HashMap<String, Command>();
+    private final Set<String> aliases;
```
Comment (deannapoh) [inline src/main/java/Command.java]: I like how you used plural form to represent multi-valued variables.

File: src/main/java/Command.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+                commandMap.put(alias, command);
+            }
+        }
+    }
+
+    Command(String... aliases) {
+        this.aliases = Set.<String>of(aliases);
+    }
+
+    public static Command fromString(String alias) {
```
Comment (deannapoh) [inline src/main/java/Command.java]: Perhaps you can rename this function so that its name is more detailed and more accurately describes what the function does?

File: src/main/java/Dash.java (hunk 2 above)
Comment (deannapoh) [inline src/main/java/Dash.java]: As these two methods have similar names despite its functionalities being different, perhaps you can rename the methods?

File: src/main/java/Dash.java [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+            botAddLine("  " + task.toString());
+            botAddLine("Now your list got " + taskList.size() + " tasks.");
+            botPrint();
+        } catch (IllegalArgumentException e) {
+            botAddLine("Your deadline must have a description and /by time!");
+            botPrint();
+        }
+    }
+
+    public static void addEvent(String msg) {
```
Comment (deannapoh) [inline src/main/java/Dash.java]: I like how you named related methods similarly. 😄

File: src/main/java/Event.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        this.to = to;
+    }
+
+    public Event(String description, boolean isDone, String from, String to) {
+        super(description, isDone);
+        this.from = from;
+        this.to = to;
+    }
+
+    public String stringify() {
```
Comment (deannapoh) [inline src/main/java/Event.java]: Perhaps you can use a more standard word to name this method?

File: src/main/java/Dash.java [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+            botAddLine(e.toString());
+            botPrint();
+        }
+    }
+
+    private static boolean hasBannedChars(String msg) {
+        return BANNED_CHARS.stream().anyMatch(msg::contains);
+    }
+
+    public static void botAddLine(String msg) {
```
Comment (deannapoh) [inline src/main/java/Dash.java]: Perhaps you can rephrase this function name so it is easier for readers to understand what it does?

**Review by deannapoh — COMMENTED** [review]
LGTM! However, there are a few improvements that can be made to the code quality, particularly the naming of methods.

File: src/main/java/dash/Dash.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+import dash.command.Command;
+import dash.exception.ExitException;
+import dash.exception.UnknownCommandException;
+import dash.task.TaskList;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.time.format.DateTimeParseException;
+import java.util.*;
```
Comment (Melvan-Chan) [inline src/main/java/dash/Dash.java]: Could we use explicit imports here?

File: src/main/java/dash/Storage.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    Storage(String filePath, TaskList taskList) {
+        this.taskList = taskList;
+        this.filePath = filePath;
+    }
+
+    private String getTaskListString() {
+        return this.taskList.stream()
+                .map(task -> task.stringify() + "\n")
+                .reduce("", (x, y) -> x + y)
+                .strip();
```
Comment (Melvan-Chan) [inline src/main/java/dash/Storage.java:50]: I like your consistent Stream indentations.

File: src/main/java/dash/Parser.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
+package dash;
+
+import dash.command.*;
```
Comment (Melvan-Chan) [inline src/main/java/dash/Parser.java]: Could we use explicit imports here as well?

File: src/main/java/dash/Storage.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
+package dash;
+
+import dash.task.*;
```
Comment (Melvan-Chan) [inline src/main/java/dash/Storage.java]: Could we use explicit imports here as well?

File: src/main/java/dash/task/Deadline.java [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+        super(description);
+        this.by = by;
+    }
+
+    public Deadline(String description, boolean isDone, LocalDate by) {
+        super(description, isDone);
+        this.by = by;
+    }
+
+    public String stringify() {
```
Comment (Melvan-Chan) [inline src/main/java/dash/task/Deadline.java:19]: Could we have a clearer method name?

File: src/main/java/dash/command/DeadlineCommand.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
+package dash.command;
+
+import dash.*;
```
Comment (Melvan-Chan) [inline src/main/java/dash/command/DeadlineCommand.java:3]: Could we use explicit imports here as well?

File: src/main/java/dash/command/EventCommand.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
+package dash.command;
+
+import dash.*;
```
Comment (Melvan-Chan) [inline src/main/java/dash/command/EventCommand.java:3]: Could we use explicit imports here as well?

**Review by Melvan-Chan — COMMENTED** [review]
Indentation discipline is great, however, some imports could be implicit.
